ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/truncate/TruncateTableAnalyzer.java

```diff
@@ -80,8 +86,25 @@ public void analyzeInternal(ASTNode root) throws SemanticException {
     checkTruncateEligibility(root, tableNode, tableNameString, table);
 
     Map<String, String> partitionSpec = getPartSpec((ASTNode) tableNode.getChild(1));
+
     addTruncateTableOutputs(tableNode, table, partitionSpec);
 
+    if (table.getStorageHandler() != null && !table.getStorageHandler().shouldTruncate(table, partitionSpec)) {
+      // Transform the truncate query to delete query.
+      StringBuilder rewrittenQuery = constructDeleteQuery(table, partitionSpec);
+      ReparseResult rr = ParseUtils.parseRewrittenQuery(conf, ctx, rewrittenQuery);
+      Context rewrittenCtx = rr.rewrittenCtx;
+      ASTNode rewrittenTree = rr.rewrittenTree;
```

Comment:
should we set the op type and prefix?
````
rewrittenCtx.setOperation(Context.Operation.DELETE);
rewrittenCtx.addDestNamePrefix(1, Context.DestClausePrefix.DELETE);
````

Reply:
Done.